SeQuant/core/eval_expr.cpp

```diff
@@ -77,64 +59,67 @@ std::string to_label_annotation(const Index& idx) {
-  auto annot = [](auto&& ixs) -> std::string {
-    using namespace ranges::views;
-
-    auto annotations = ixs | transform(to_label_annotation);
+std::string EvalExpr::indices_annot() const noexcept {
+  using ranges::views::filter;
 
-    return annotations                      //
-           | intersperse(std::string{","})  //
-           | join                           //
-           | ranges::to<std::string>;
-  };
+  if (!is_tensor()) return {};
 
-  auto nested = NestedTensorIndices{as_tensor()};
+  auto outer = csv_labels(canon_indices_ |
+                          filter(ranges::not_fn(&Index::has_proto_indices)));
+  auto inner = csv_labels(canon_indices_ | filter(&Index::has_proto_indices));
 
-  return nested.inner.empty()  //
-             ? annot(nested.outer)
-             : annot(nested.outer) + ";" + annot(nested.inner);
+  return outer + (inner.empty() ? "" : (";" + inner));
 }
 
-size_t EvalExpr::global_id_{};
+EvalExpr::index_vector const& EvalExpr::canon_indices() const noexcept {
+  return canon_indices_;
+}
 
 EvalExpr::EvalExpr(Tensor const& tnsr)
     : op_type_{EvalOp::Id},
       result_type_{ResultType::Tensor},
-      hash_value_{hash_terminal_tensor(tnsr)},
-      id_{},
-      expr_{tnsr.clone()},
-      tot_{is_tot(tnsr)} {}
+      expr_{tnsr.clone()} {
+  ExprPtrList tlist{expr_};
+  auto tn = TensorNetwork(tlist);
+  auto md =
+      tn.canonicalize_slots(TensorCanonicalizer::cardinal_tensor_labels());
+  hash_value_ = md.hash_value();
+  canon_indices_ = md.get_indices() | ranges::to<index_vector>;
+  canon_phase_ = md.phase;
+}
 
 EvalExpr::EvalExpr(Constant const& c)
     : op_type_{EvalOp::Id},
       result_type_{ResultType::Scalar},
       hash_value_{hash::value(c)},
-      id_{},
-      expr_{c.clone()},
-      tot_{false} {}
+      expr_{c.clone()} {}
 
 EvalExpr::EvalExpr(Variable const& v)
     : op_type_{EvalOp::Id},
       result_type_{ResultType::Scalar},
       hash_value_{hash::value(v)},
-      id_{},
-      expr_{v.clone()},
-      tot_{false} {}
+      expr_{v.clone()} {}
+
+EvalExpr::EvalExpr(EvalOp op, ResultType res, ExprPtr const& ex,
+                   index_vector ixs, std::int8_t p, size_t h)
+    : op_type_{op},
+      result_type_{res},
+      expr_{ex.clone()},
+      canon_indices_{std::move(ixs)},
+      canon_phase_{p},
+      hash_value_{h} {}
 
 EvalExpr::EvalExpr(EvalExpr const& left, EvalExpr const& right, EvalOp op)
     : op_type_{op},
       hash_value_{hash_imed(left, right, op)},
-      id_{++global_id_},
       expr_{make_imed(left, right, op)} {
   result_type_ = expr_->is<Tensor>() ? ResultType::Tensor : ResultType::Scalar;
-  tot_ = expr_->is<Tensor>() && is_tot(expr_->as<Tensor>());
+  if (result_type() == ResultType::Tensor) {
+    auto tn = TensorNetwork(expr_);
+    auto canon =
+        tn.canonicalize_slots(TensorCanonicalizer::cardinal_tensor_labels());
+    canon_indices_ = canon.get_indices() | ranges::to<index_vector>;
+  }
```

Comment:
@bimalgaudel does this need to adjust for the phase also?

Reply:
Sure. But this function is deprecated and will be removed soon.